NTPClient.cpp

```diff
@@ -149,6 +149,97 @@ int NTPClient::getSeconds() const {
   return (this->getEpochTime() % 60);
 }
 
+int NTPClient::getDay() const {
+
+  long days = this->getEpochTime()  / 86400L;
+  int fullYears = days / 365;
+  int overDays = days % 365;
+
+  int leapYears = (fullYears - 2) / 4;
+  if (leapYears > overDays) {
+	fullYears--;
+  }
+
+  int currentYear = 1970 + fullYears;
+
+  int thisYearIsLeap = currentYear % 4 == 0 ? 1 : 0;
+
+  int dayOfYear = (days - leapYears) % ( 365 + thisYearIsLeap);
+  if(dayOfYear == 0) {
+	dayOfYear = 365 + thisYearIsLeap;
+  }
+
+  int daysInMonth[12] = {31, 28 + thisYearIsLeap, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+
+  for( int month = 0; month < 12; month++) {
+	if(dayOfYear < daysInMonth[month]) {
+		return dayOfYear;
+	} else {
+		dayOfYear -= daysInMonth[month];
+	}
+  }
+
+  return -1;
+}
+
+int NTPClient::getMonth() const {
+
+  long days = this->getEpochTime()  / 86400L;
+  int fullYears = days / 365;
+  int overDays = days % 365;
+
+  int leapYears = (fullYears - 2) / 4;
+  if (leapYears > overDays) {
+	fullYears--;
+  }
+
+  int currentYear = 1970 + leapYears;
+
+  int thisYearIsLeap = currentYear % 4 == 0 ? 1 : 0;
+
+  int dayOfYear = (days - leapYears) % ( 365 + thisYearIsLeap);
+  if(dayOfYear == 0) {
+	dayOfYear = 365 + thisYearIsLeap;
+  }
+
+  int daysInMonth[12] = {31, 28 + thisYearIsLeap, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+
+  for( int month = 0; month < 12; month++) {
+	if(dayOfYear < daysInMonth[month]) {
+		return month + 1;
+	} else {
+		dayOfYear -= daysInMonth[month];
+	}
+  }
+
+  return -1;
+}
+
+int NTPClient::getYear() const {
+  long days = this->getEpochTime()  / 86400L;
+  int fullYears = days / 365;
+  int overDays = days % 365;
+
+  int leapYears = (fullYears - 2) / 4;
```

Comment:
`getYear()` fails for all days that match December 31st, except for the year 1970 and those where the next year is a leap one:

| Code | 1970/12/31 | 1971/12/31 | 1972/12/31 |
| - | - | - | - |
| `days` | 364 | 729 _(365+364)_ | 1095 _(365+365+365)_ |
| `fullYears = days / 365` | 0 _(364 / 365)_ | 1 _(729 / 365)_ | 3 _(1095 / 365)_ |
| `overDays = days % 365` | 364 _(364 % 365)_ | 364 _(729 % 365)_ | 0 _(1095 % 365)_ |
| `leapYears = (fullYears - 2) / 4` | 0 _((0 − 2) / 4)_ | 0 _((1 − 2) / 4)_ | 0 _((3 − 2) / 4)_ |
| `if (leapYears > overDays) fullYears--` | 0 _(No change)_ | 1 _(No change)_ | 3 _(No change)_ |
| `return 1970 + fullYears` | **1970** _(1970 + 0)_ | **1971** _(1970 + 1)_ | **1973** _(1970 + 3)_ |

Fortunately, it can be easily fixed:

```suggestion
  int leapYears = (fullYears + 1) / 4;
```